# Matriks invers



Matriks yang jika dikalikan dengan matriks asalnya, akan menghasilkan metriks identitas.
Dan sering digunakan dalam pemecahan sistem persamaan linier.

AA^-1 = A^-1 x A = I

I adalah matriks identitas yang memiliki elemen-elemen diagonal bernilai 1 dan elemen-elemen lainnya bernilai 0. Matriks invers ini hanya ada untuk matriks persegi yang tidak singular, determinannya tidak nol. Contoh :

- Matriks dengan ordo(2 x2):


<!-- \begin{equation*} -->
A=
\begin{bmatrix}
a&b\\
c&d
\end{bmatrix}
<!-- \end{equation*} -->

\begin{align*}
A^-1\ =\frac{1}{ad-bc}
\begin{bmatrix}
d&-b\\
-c&a
\end{bmatrix}
\end{align*}


## **Sifat-sifat matriks**  
JIka A mempunyai matrik invers, k bilangan bulat positif, dan c adalah skalar $\neq 0$ maka
1. $A^{-1}$ diinverskan menjadi $(A^{-1})^{-1}=A$
2. $A^{k}$ diinverskan menjadi $(A^{k})^{-1}=A^{k} A^{k} A^{k}=(A^{-1})^{k}=A^{-k}$
3. $cA$ diinverskan dan $(cA)^{-1}=\frac{1}{c}  A^{-1}, c\neq 0$
4. $A^{T}$ diinverskan dan $(A^{T})^{-1}=(A^{-1})^{T}$

### Penyelsaian sistem persamaan linier mengunakan matriks invers
membuat 4 soal persamaan dan 4 variabel dan cari solusinya pakai invers matriks
algoritma eliminasi gauss

Penyelesaian 4 persamaan linier menggunakan matrik invers.
$Ax=b\\
x=A^{-1}b
$

\begin{align*}
\\3x_1+x_2-2x_3-3x_4=4\\
x_1+x_2-x_3-2x_4=-2\\
-4x_1-x_2+3x_3-3x_4=3\\
3x_1+x_2-2x_3+x_4=6\\
\end{align*}
A=
\begin{bmatrix}
3&1&-2&-3\\
1&1&-1&-2\\
-4&-1&3&-3\\
3&1&-2&1\\
\end{bmatrix}

X=
\begin{bmatrix}
x_1\\
x_2\\
x_3\\
x_4\\
\end{bmatrix}
B=
\begin{bmatrix}
4\\
-2\\
3\\
6\\
\end{bmatrix}
$



In [1]:
def RowSwap(A,k,l):
# =============================================================================
#     A is a NumPy array.  RowSwap will return duplicate array with rows
#     k and l swapped.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A

    B = np.copy(A).astype('float64')

    for j in range(n):
        temp = B[k][j]
        B[k][j] = B[l][j]
        B[l][j] = temp

    return B

def RowScale(A,k,scale):
# =============================================================================
#     A is a NumPy array.  RowScale will return duplicate array with the
#     entries of row k multiplied by scale.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[k][j] *= scale

    return B

def RowAdd(A,k,l,scale):
# =============================================================================
#     A is a numpy array.  RowAdd will return duplicate array with row
#     l modifed.  The new values will be the old values of row l added to
#     the values of row k, multiplied by scale.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[l][j] += B[k][j]*scale

    return B

In [2]:
import numpy as np
A = np.array([[3,1,-2,-3,1,0,0,0], [1,1,-1,-2,0,1,0,0], [-4,-1,3,-3,0,0,1,0], [3,1,-2,1,0,0,0,1]])
print(A)

[[ 3  1 -2 -3  1  0  0  0]
 [ 1  1 -1 -2  0  1  0  0]
 [-4 -1  3 -3  0  0  1  0]
 [ 3  1 -2  1  0  0  0  1]]


In [3]:
B = RowSwap(A, 0, 1)
print(B)

[[ 1.  1. -1. -2.  0.  1.  0.  0.]
 [ 3.  1. -2. -3.  1.  0.  0.  0.]
 [-4. -1.  3. -3.  0.  0.  1.  0.]
 [ 3.  1. -2.  1.  0.  0.  0.  1.]]


In [4]:
C = RowAdd(B, 1, 3, -1)
print(C)

[[ 1.  1. -1. -2.  0.  1.  0.  0.]
 [ 3.  1. -2. -3.  1.  0.  0.  0.]
 [-4. -1.  3. -3.  0.  0.  1.  0.]
 [ 0.  0.  0.  4. -1.  0.  0.  1.]]


In [5]:
D = RowAdd(C, 0, 2, 4)
print(D)

[[  1.   1.  -1.  -2.   0.   1.   0.   0.]
 [  3.   1.  -2.  -3.   1.   0.   0.   0.]
 [  0.   3.  -1. -11.   0.   4.   1.   0.]
 [  0.   0.   0.   4.  -1.   0.   0.   1.]]


In [6]:
E = RowAdd(D, 0, 1, -3)
print(E)

[[  1.   1.  -1.  -2.   0.   1.   0.   0.]
 [  0.  -2.   1.   3.   1.  -3.   0.   0.]
 [  0.   3.  -1. -11.   0.   4.   1.   0.]
 [  0.   0.   0.   4.  -1.   0.   0.   1.]]


In [7]:
F = RowScale(E, 1, -1/2)
print(F)

[[  1.    1.   -1.   -2.    0.    1.    0.    0. ]
 [ -0.    1.   -0.5  -1.5  -0.5   1.5  -0.   -0. ]
 [  0.    3.   -1.  -11.    0.    4.    1.    0. ]
 [  0.    0.    0.    4.   -1.    0.    0.    1. ]]


In [8]:
G = RowAdd(F, 1, 2, -3)
print(G)

[[ 1.   1.  -1.  -2.   0.   1.   0.   0. ]
 [-0.   1.  -0.5 -1.5 -0.5  1.5 -0.  -0. ]
 [ 0.   0.   0.5 -6.5  1.5 -0.5  1.   0. ]
 [ 0.   0.   0.   4.  -1.   0.   0.   1. ]]


In [9]:
H = RowAdd(G, 2, 1, 1)
print(H)

[[ 1.   1.  -1.  -2.   0.   1.   0.   0. ]
 [ 0.   1.   0.  -8.   1.   1.   1.   0. ]
 [ 0.   0.   0.5 -6.5  1.5 -0.5  1.   0. ]
 [ 0.   0.   0.   4.  -1.   0.   0.   1. ]]


In [10]:
I = RowScale(H, 2, 2)
print(I)

[[  1.   1.  -1.  -2.   0.   1.   0.   0.]
 [  0.   1.   0.  -8.   1.   1.   1.   0.]
 [  0.   0.   1. -13.   3.  -1.   2.   0.]
 [  0.   0.   0.   4.  -1.   0.   0.   1.]]


In [11]:
J = RowScale(I, 3, 1/4)
print(J)

[[  1.     1.    -1.    -2.     0.     1.     0.     0.  ]
 [  0.     1.     0.    -8.     1.     1.     1.     0.  ]
 [  0.     0.     1.   -13.     3.    -1.     2.     0.  ]
 [  0.     0.     0.     1.    -0.25   0.     0.     0.25]]


In [12]:
K = RowAdd(J, 3, 2, 13)
print(K)

[[ 1.    1.   -1.   -2.    0.    1.    0.    0.  ]
 [ 0.    1.    0.   -8.    1.    1.    1.    0.  ]
 [ 0.    0.    1.    0.   -0.25 -1.    2.    3.25]
 [ 0.    0.    0.    1.   -0.25  0.    0.    0.25]]


In [13]:
L = RowAdd(K, 3, 1, 8)
print(L)

[[ 1.    1.   -1.   -2.    0.    1.    0.    0.  ]
 [ 0.    1.    0.    0.   -1.    1.    1.    2.  ]
 [ 0.    0.    1.    0.   -0.25 -1.    2.    3.25]
 [ 0.    0.    0.    1.   -0.25  0.    0.    0.25]]


In [14]:
M = RowAdd(L, 3, 0, 2)
print(M)

[[ 1.    1.   -1.    0.   -0.5   1.    0.    0.5 ]
 [ 0.    1.    0.    0.   -1.    1.    1.    2.  ]
 [ 0.    0.    1.    0.   -0.25 -1.    2.    3.25]
 [ 0.    0.    0.    1.   -0.25  0.    0.    0.25]]


In [15]:
N = RowAdd(M, 2, 0, 1)
print(N)

[[ 1.    1.    0.    0.   -0.75  0.    2.    3.75]
 [ 0.    1.    0.    0.   -1.    1.    1.    2.  ]
 [ 0.    0.    1.    0.   -0.25 -1.    2.    3.25]
 [ 0.    0.    0.    1.   -0.25  0.    0.    0.25]]


In [16]:
O = RowAdd(N,1,0,-1)
print(O)

[[ 1.    0.    0.    0.    0.25 -1.    1.    1.75]
 [ 0.    1.    0.    0.   -1.    1.    1.    2.  ]
 [ 0.    0.    1.    0.   -0.25 -1.    2.    3.25]
 [ 0.    0.    0.    1.   -0.25  0.    0.    0.25]]


In [17]:
P = np.array([[0.25,-1,1,1.75], [-1,1,1,2], [-0.25,-1,2,3.25], [-0.25,0,0,0.25]])
Q = np.array([[4], [-2], [3], [6]])
R = np.array([[3,1,-2,-3], [1,1,-1,-2], [-4,-1,3,-3], [3,1,-2,1]])
print(P@Q)

[[16.5]
 [ 9. ]
 [26.5]
 [ 0.5]]



Matriks Inverse=
\begin{bmatrix}
0.25&-1&1&1.75\\
-1&1&1&2\\
-0.25&-1&2&3.25\\
-0.25&0&0&0.25\\
\end{bmatrix}

X=
\begin{bmatrix}
16.5\\
9\\
26.5\\
0.5\\
\end{bmatrix}
B=
\begin{bmatrix}
4\\
-2\\
3\\
6\\
\end{bmatrix}






\begin{align*}
\\x_1+x_2=3\\
2x_1+4x_2=8\\
\end{align*}
A=
\begin{bmatrix}
1&1\\
2&4\\
\end{bmatrix}

X=
\begin{bmatrix}
x_1\\
x_2\\
\end{bmatrix}

B=
\begin{bmatrix}
3\\
8\\
\end{bmatrix}



In [18]:
def RowSwap(A,k,l):
# =============================================================================
#     A is a NumPy array.  RowSwap will return duplicate array with rows
#     k and l swapped.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A

    B = np.copy(A).astype('float64')

    for j in range(n):
        temp = B[k][j]
        B[k][j] = B[l][j]
        B[l][j] = temp

    return B

def RowScale(A,k,scale):
# =============================================================================
#     A is a NumPy array.  RowScale will return duplicate array with the
#     entries of row k multiplied by scale.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[k][j] *= scale

    return B

def RowAdd(A,k,l,scale):
# =============================================================================
#     A is a numpy array.  RowAdd will return duplicate array with row
#     l modifed.  The new values will be the old values of row l added to
#     the values of row k, multiplied by scale.
# =============================================================================
    m = A.shape[0]  # m is number of rows in A
    n = A.shape[1]  # n is number of columns in A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[l][j] += B[k][j]*scale

    return B

In [19]:
import numpy as np
A = np.array([[1,1,1,0],[2,4,0,1]])
print(A)

[[1 1 1 0]
 [2 4 0 1]]


In [20]:
B = RowAdd (A, 0, 1, -2)
print(B)

[[ 1.  1.  1.  0.]
 [ 0.  2. -2.  1.]]


In [21]:
C = RowScale (B, 1, 1/2)
print(C)

[[ 1.   1.   1.   0. ]
 [ 0.   1.  -1.   0.5]]


In [22]:
D = RowAdd (C, 1, 0, -1)
print(D)

[[ 1.   0.   2.  -0.5]
 [ 0.   1.  -1.   0.5]]


In [23]:
E = np.array([[2, -0.5], [-1,0.5]])
F = np.array([[3], [8]])
G = np.array([[1,1],[2,4]])
print(E@F)

[[2.]
 [1.]]



Matriks Inverse=
\begin{bmatrix}
2&-0.5\\
-1&0.5\\
\end{bmatrix}

X=
\begin{bmatrix}
2\\
1\\
\end{bmatrix}
B=
\begin{bmatrix}
3\\
8\\
\end{bmatrix}




In [24]:
Q1= np.array([[3,1,-2,-3]])
Q2 = np.dot(P,Q)
print(Q1@Q2)

[[4.]]
